pilot/pkg/networking/core/listener.go

```diff
@@ -498,16 +498,25 @@ func (lb *ListenerBuilder) buildSidecarOutboundListeners(node *model.Proxy,
 							// Make sure each endpoint address is a valid address
 							// as service entries could have NONE resolution with label selectors for workload
 							// entries (which could technically have hostnames).
-							if !netutil.IsValidIPAddress(instance.Address) {
+							isValidInstance := true
+							for _, addr := range instance.Addresses {
+								if !netutil.IsValidIPAddress(addr) {
+									isValidInstance = false
+									break
+								}
+							}
+							if !isValidInstance {
 								continue
 							}
 							// Skip build outbound listener to the node itself,
 							// as when app access itself by pod ip will not flow through this listener.
 							// Simultaneously, it will be duplicate with inbound listener.
-							if instance.Address == node.IPAddresses[0] {
+							// should continue if current IstioEndpoint instance has the same ip with the
+							// first ip of node IPaddresses
+							if instance.FirstAddressOrNil() == node.IPAddresses[0] {
```

Comment:
How do we know the first will match in both of these lists? If we do know, we should indicate on the comment why its gauranteed

Reply:
The reason why the comparison can be performed is because the first entry in the addresses list of both objects comes from k8s podIPs.
In case of the proxy object, the first IP is obtained using INSTANCE_IP env + downward API, while for the istio endpoint the addresses comes from parsing the k8s pod object.

I have added a comment to the conditional checks to clarify this.